cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/store/NamespaceSourceControlMetadataStore.java

```diff
@@ -0,0 +1,219 @@
+    SourceControlMeta existingSourceControlMeta = get(appRef);
+    if (sourceControlMeta != null && existingSourceControlMeta != null
+        && sourceControlMeta.getLastSyncedAt()
+        .isBefore(existingSourceControlMeta.getLastSyncedAt())) {
+      List<Field<?>> fields = new ArrayList<>();
+      fields.add(
+          Fields.longField(StoreDefinition.NamespaceSourceControlMetadataStore.LAST_MODIFIED_FIELD,
+              sourceControlMeta.getLastSyncedAt().toEpochMilli()));
+      throw new InvalidFieldException(
+          StoreDefinition.NamespaceSourceControlMetadataStore.NAMESPACE_SOURCE_CONTROL_METADATA,
+          fields, "An updated row having a greater last modified is already present");
+    }
+    scmTable.upsert(getAllFields(appRef, sourceControlMeta, existingSourceControlMeta));
+  }
+
+  /**
+   * Deletes the source control metadata associated with the specified application ID from the
+   * namespace.
+   *
+   * @param appRef {@link ApplicationReference} whose source control metadata is to be deleted.
+   * @throws IOException if it failed to read or delete the metadata
+   */
+  public void delete(ApplicationReference appRef) throws IOException {
+    getNamespaceSourceControlMetadataTable().delete(getPrimaryKey(appRef));
+  }
+
+  /**
+   * Deletes all rows of source control metadata within the specified namespace.
+   *
+   * @param namespace The namespace for which all source control metadata rows are to be deleted.
+   * @throws IOException if it failed to read or delete the metadata.
+   */
+  public void deleteAll(String namespace) throws IOException {
+    getNamespaceSourceControlMetadataTable().deleteAll(getNamespaceRange(namespace));
+  }
+
+  private Collection<Field<?>> getAllFields(ApplicationReference appRef,
+      SourceControlMeta newSourceControlMeta, SourceControlMeta existingSourceControlMeta) {
+    List<Field<?>> fields = getPrimaryKey(appRef);
+    fields.add(Fields.stringField(
+        StoreDefinition.NamespaceSourceControlMetadataStore.SPECIFICATION_HASH_FIELD,
+        newSourceControlMeta == null ? null : newSourceControlMeta.getFileHash()));
+    fields.add(
+        Fields.stringField(StoreDefinition.NamespaceSourceControlMetadataStore.COMMIT_ID_FIELD,
+            newSourceControlMeta == null ? null : newSourceControlMeta.getCommitId()));
+    // Whenever an app is deployed, the expected behavior is that the last modified field will be
+    // retained and not reset.
+    fields.add(
+        Fields.longField(StoreDefinition.NamespaceSourceControlMetadataStore.LAST_MODIFIED_FIELD,
+            newSourceControlMeta == null ? (existingSourceControlMeta == null ? 0L
+                : existingSourceControlMeta.getLastSyncedAt().toEpochMilli())
+                : newSourceControlMeta.getLastSyncedAt().toEpochMilli()));
+    fields.add(
+        Fields.booleanField(StoreDefinition.NamespaceSourceControlMetadataStore.IS_SYNCED_FIELD,
```

Comment:
hmm now that I think about it you can always send a sourcecontrol meta. just the sycned field will be false. That is the base state when a app is added. That way you do not have to deal with null source control meta at all.

Reply:
Right. Done. Thank you for pointing that out